File: @FileManager/EpicToolbox.m
function allTrialsData=EpicToolbox(obj, fileList)
% For an list of files from sf_post combine all the sensors
% of the same trial into a mat file matching the structure for usage with
% EpicToolbox. Intended to retrieve data in small chunks without writing
% files.
%
% trialData=EpicToolbox(fileList)

%% Transform data for EpicToolbox

%% Combine data from different sensors into one struct

pathfields=obj.folderLevels;

sensorLevel='Sensor';
if ~ismember('Sensor',pathfields)
    sensorLevel='sensor';
    if ~ismember('sensor',pathfields)
        error('FileManager object must contain Sensor folderLevel in pathStructure');
    end
end

if isempty(fileList)
    warning('No files to create an EpicToolbox struct');
    allTrialsData={};
    return;
end

%Last field is always trialnames which we certainly need to group
% Also group Subject and Date fields if they exist.
% Sensor field should allways be present to be combined as topics in the
% trial.

lastfield=pathfields{end}; 

% tolook represents fields that I want to look for in order to group files
% that belong to the same trial and thus should be together.

% tolook={'Subject','Date','Ambulation'}; % To reduce the ouput to only
% certain fields.

tolook=setdiff(pathfields(1:end-1),sensorLevel); % To keep the output to all the fields (except sensor and last field)

trials=obj.getFields(fileList,lastfield);
sensors=obj.getFields(fileList,sensorLevel);
individualtrialstbl=cell2table(trials,'VariableNames',{lastfield});

for i=1:numel(tolook)
    fieldname=tolook{i};
    if any(contains(pathfields,fieldname))
        a=obj.getFields(fileList,fieldname);
        atbl=cell2table(a,'VariableNames',{fieldname});
        individualtrialstbl=horzcat(individualtrialstbl,atbl);
    end    
end

[unique_trials,uniqueTrialIdx,trialIdx]=unique(individualtrialstbl);

allTrialsData=cell(size(unique_trials,1),1);

%%
for i=1:numel(fileList)
    % Recover trialData stored in allTrialls cell
    trialData=allTrialsData{trialIdx(i)};
               
    %% Load data for each sensor into trialData struct        
    sensor=sensors{i};
    if ~exist(fileList{i},'file')
        continue;
    end
    a=load(fileList{i});
    if isfield(trialData,sensor)
        error('Trial already contains %s data\n @file:%s\n',sensor,fileList{i});
    end
    if ((numel(fieldnames(a))==1) && isfield(a,'data'))
        eval(sprintf('trialData.%s=a.data;',sensor));
    else
        eval(sprintf('trialData.%s=a;',sensor));
    end
    
    allTrialsData{trialIdx(i)}=trialData;
end
% After loading data fill out info into every trial to give user more info
% about each trial.
for i=1:height(unique_trials)    
    a=allTrialsData{i};    
    varnames=individualtrialstbl.Properties.VariableNames;
    for j=1:size(individualtrialstbl,2)
        a.info.(varnames{j})=unique_trials{i,j}{:};
    end
    allTrialsData{i}=a;
end
    
    
    
    
    
    
    
    
    
    
    
    
    
    
  
  

File: python/EpicToolbox/filemanager/_epictoolbox.py
import pandas as pd
import h5py
import numpy as np
from joblib import Parallel, delayed
import os

def loadhdf(hdf5file,dataframe=True,verbose=0):
    '''Read data stored on hdf5 file data should
    contain a key 'data' and a field 'names' for the header'''

    if verbose>1:
        print("Loading {}".format(hdf5file))
    with h5py.File(hdf5file, 'r') as f:
        #Read data and names and construct only dataframe
        out_data=f['data'][:]
        out_names_refs=f['names'][:]
        a=[f[obj[0]][:] for obj in out_names_refs]        
        names=[''.join([chr(c[0]) for c in b]) for b in a] # python 3.8
        #names=[''.join(chr(c) for c in f[obj[0]]) for obj in out_names_refs]

        if dataframe:
            out=pd.DataFrame(data=np.transpose(out_data),columns=names)
        else:
            out=out_data
    return out


def EpicToolbox(self,fileList,n_jobs=1):
    '''For an list of files from sf_post combine all the sensors
    of the same trial into a mat file matching the structure for usage with
    EpicToolbox. Intended to retrieve data in small chunks without writing
    files.

    trialData=EpicToolbox(fileList)
    Transform data for EpicToolbox

    Combine data from different sensors into one struct
    '''

    if self.verbose>0:
        print("Loading {} files to EpicToolbox".format(len(fileList)))

    pathfields=self.folderLevels
    lastfield=self.folderLevels[-1]

    tolook=[level for level in pathfields[0:-1] if ((level != 'Sensor') and (level != 'sensor'))]

    trials=self.getFields(fileList,lastfield)

    if 'Sensor' in pathfields:
        sensors=self.getFields(fileList,'Sensor')
    else:
        sensors=self.getFields(fileList,'sensor')

    individualtrialstbl=pd.DataFrame()
    for field in tolook:
        if field in pathfields:
            vals=pd.Series(self.getFields(fileList,field),name=field)
            individualtrialstbl=pd.concat([individualtrialstbl,vals],axis=1)

    individualtrialstbl=pd.concat([individualtrialstbl,pd.Series(trials,name=lastfield)],axis=1)
    #individualtrialstbl.drop_duplicates()

    combined=individualtrialstbl.apply(lambda x: os.path.sep.join(x), axis=1)
    uniquetrials,uniqueidx=np.unique(combined,return_inverse=True)

    trialdata=[None]*len(uniquetrials)

    alldfs=Parallel(n_jobs=n_jobs)(delayed(loadhdf)(file,verbose=self.verbose) for file in fileList)
    #alldfs=[loadhdf(file) for file in fileList]

    if self.verbose>0:
        print('Finished loading')

    for i,df in enumerate(alldfs):
        if trialdata[uniqueidx[i]] is None:
            trialdata[uniqueidx[i]]=dict()
        trialdata[uniqueidx[i]][sensors[i]]=df.copy()
        trialdata[uniqueidx[i]]['info']=individualtrialstbl.loc[uniqueidx[i],:].copy()

    return trialdata
